dwave/preprocessing/include/dwave/presolve.h

```diff
@@ -0,0 +1,410 @@
+
+                // ax ◯ c
+                bias_type a = constraint.linear(v);
+                assert(a);  // should have already been removed if 0
+
+                // offset should have already been removed but may as well be safe
+                bias_type rhs = (constraint.rhs() - constraint.offset()) / a;
+
+                // todo: test if negative
+
+                if (constraint.sense() == dimod::Sense::EQ) {
+                    model_.set_lower_bound(v, std::max(rhs, model_.lower_bound(v)));
+                    model_.set_upper_bound(v, std::min(rhs, model_.upper_bound(v)));
+                } else if ((constraint.sense() == dimod::Sense::LE) != (a < 0)) {
+                    model_.set_upper_bound(v, std::min(rhs, model_.upper_bound(v)));
+                } else {
+                    assert((constraint.sense() == dimod::Sense::GE) == (a >= 0));
+                    model_.set_lower_bound(v, std::max(rhs, model_.lower_bound(v)));
+                }
+
+                model_.remove_constraint(c);
+                changes = true;
+                continue;
+            }
+
+            ++c;
+        }
+
+        // *-- tighten bounds based on vartype
+        bias_type lb;
+        bias_type ub;
+        for (size_type v = 0; v < model_.num_variables(); ++v) {
+            switch (model_.vartype(v)) {
+                case dimod::Vartype::SPIN:
+                case dimod::Vartype::BINARY:
+                case dimod::Vartype::INTEGER:
+                    ub = model_.upper_bound(v);
+                    if (ub != std::floor(ub)) {
+                        model_.set_upper_bound(v, std::floor(ub));
+                        changes = true;
+                    }
+                    lb = model_.lower_bound(v);
+                    if (lb != std::ceil(lb)) {
+                        model_.set_lower_bound(v, std::ceil(lb));
+                        changes = true;
+                    }
+                    break;
+                case dimod::Vartype::REAL:
+                    break;
+            }
+        }
+
+        // *-- remove variables that are fixed by bounds
+        size_type v = 0;
+        while (v < model_.num_variables()) {
+            if (model_.lower_bound(v) == model_.upper_bound(v)) {
+                postsolver_.fix_variable(v, model_.lower_bound(v));
+                model_.fix_variable(v, model_.lower_bound(v));
+            }
+            ++v;
+        }
+    }
+
+    // Cleanup
+
+    // *-- remove any invalid discrete markers
+    std::vector<index_type> discrete;
+    for (size_type c = 0; c < model_.num_constraints(); ++c) {
+        auto& constraint = model_.constraint_ref(c);
+
+        if (!constraint.marked_discrete()) continue;
+
+        // we can check if it's well formed
+        if (constraint.is_onehot()) {
+            discrete.push_back(c);
+        } else {
+            constraint.mark_discrete(false);  // if it's not one-hot, it's not discrete
```

Comment:
So this will essentially silently "fix" incorrectly marked constraints?

Reply:
yes